text/0471-on-modifier.md

```diff
@@ -0,0 +1,216 @@
+  method). This is problematic behavior, not just because of the transform, but
+  because of the implicit state that is being passed to the modifier in the
+  first place, and the fact that the modifier is "magical" in the sense that it
+  can do things that no other modifier or helper can do.
+- Binds to the `click` event by default, which requires a bit of knowledge to
+  understand on first glance, and requires a bit of extra knowledge to figure
+  out how to bind to other events. Compared to `onclick=`, this is less
+  explicit, and requires more knowledge of the framework out of the box.
+- Is confusing when used with the `@action` decorator to bind the function, and
+  the `{{action}}` helper. The term "action" is highly overloaded here, and it
+  is unclear what it concretely means without context.
+
+The `on*` properties, by contrast, may seem to be a better approach overall to
+the problem, but they have their own issues:
+
+- The require the ability to bind to element _properties_. Properties are _not_
+  attributes, and the distinction is subtle, but important. Most importantly,
+  props are _not_ rendered, and thus are _not_ SSR/rehydration friendly, since
+  it is difficult to assign during rehydration. By constrast, modifiers are
+  specifically designed around SSR. They do not trigger at all during server
+  side render, and then are run specifically during rehydration.
+- Not all events are bindable via `on*` properties, and some events such as
+  `focus` have different behavior than when assigned with `addEventListener`
+  directly. This behavior has been a pain point in other frontend frameworks,
+  and requires workarounds in most cases.
+- They do not work on certain elements at all, such as `<svg>`.
+- They are not compatible at all with standard web components. Web components
+  don't send events in the same way as standard elements for properties, and
+  instead rely on event listeners added via `addEventListener`. While web
+  components are not currently used in the Ember ecosystem, they are advancing
+  and may eventually be useful for smaller, self-contained components,
+  especially for users that want to share code between framework ecosystems.
+
+Neither of these solutions is perfect for solving the problem of adding event
+listeners. This RFC proposes adding a new modifier to Ember: the `{{on}}`
+modifier. This modifier will explicitly add event listeners using the
+`addEventListener` API.
+
+## Credits
+
+The design of this modifier is based on [ember-on-modifier](https://github.com/buschtoens/ember-on-modifier)
+by Jan Buschtöns, which is an excellent addon that has allowed us to test this
+design in real apps and get feedback about the design.
+
+## Detailed design
+
+The `{{on}}` modifier will recieve:
+
+1. The event name as a string as the first positional parameter
+2. The event listener function as the second positional parameter
+3. Named parameters as options
```

Comment:
This is kinda late and I apologize for that, but what will be the behavior, if the event name or event listener are missing or have the wrong type. `ember-on-modifier` currently checks for `typeof eventName === 'string' && typeof callback === 'function'` and only then registers the listener. If the check fails, no error will be raised. We might want to make this a hard assertion instead.

Reply:
Agreed, I believe that we will validate name and listener via assertions. I suppose we could do something to allow null/undefined but it really seems more likely to be a bug than intentional.

Reply:
@richard-viney One solution would be using `(optional)` from `ember-composable-helpers`, which returns a no-op function, if the input is not a function:

```hbs
<button {{on "click" (optional (if this.isListening this.listener))}}>
  ...
</button>
```

While this requires extra gymnastics, I agree with @pzuraq and @rwjblue, that it is probably safer to assert against missing `eventName` and `callback` to prevent accidental errors.